### ISSUE TITLE
[Enhancement]  Rewrite Plugin should rewrite http to https

### ISSUE THREAD

**Issue by schmitch** [mailgun/vulcand#120]
Hello,
It would be cool, if vulcand could have a native way to redirect http to https.
Currently the rewrite plugin only allows to redirect a location/path to another.
It would be cool to have a Matcher that allows to match the complete URL and redirect to another like http://foo.bar to https://foo.bar or like http://foo.bar/admin/* to https://foor.bar

**Comment by porjo** [comment]
I've played with this again using v0.8.0, unfortunately I still can't get it to work.

The rewrite middleware does not receive the `https` scheme in the URL [here](https://github.com/mailgun/vulcand/blob/master/plugin/rewrite/rewrite.go#L152). Because of this, the input URL always defaults to `http` and so the rewrite is applied every time, resulting in a redirect loop.

**Comment by klizhentas** [comment]
OK, thanks for your research. Not sure why https is not passed though, needs more research

**Comment by porjo** [comment]
According to the docs for [http.Request](http://golang.org/pkg/net/http/#Request), referencing `request.URL.Scheme` (as is being done in rewrite.go) is not expected to work:

> For server requests the URL is parsed from the URI
> supplied on the Request-Line as stored in RequestURI.  For
> most requests, fields other than Path and RawQuery will be
> empty. (See RFC 2616, Section 5.1.2)

It seems that checking `request.TLS != nil` is the standard way of determining scheme in this case, however that doesn't work in rewrite.go either as the value is always `nil`. I suspect that something in the *manners* lib may need to copy this over.

**Comment by klizhentas** [comment]
Great, thanks for digging into this. I assume that this function is the one we need to fix:

https://github.com/mailgun/vulcand/blob/master/Godeps/_workspace/src/github.com/mailgun/oxy/stream/stream.go#L278

This is the vendored oxy lib that forgets to copy over the flag

**Comment by porjo** [comment]
> This is the vendored oxy lib that forgets to copy over the flag

Unfortunately that's not it. I've just tested and `req.TLS` is always nil in that function (even for https requests). There must be another point further upstream where it's getting stripped.

**Comment by klizhentas** [comment]
If you are curious, here was the bug and the resolution:

https://github.com/mailgun/manners/pull/4

I will merge this in this week

**Comment by porjo** [comment]
@klizhentas Thanks - I look forward to the fix.

**Comment by klizhentas** [comment]
merged, *TLS is not nil now, however I did not change logic inside `rewrite` yet

**Comment by klizhentas** [comment]
Fixed

**Comment by porjo** [comment]
@klizhentas thanks

@schmitch FYI, the following rewrite rule is working for me:

```json
etcdctl set /vulcand/frontends/test1/middlewares/r1 \
'{"Type": "rewrite", "Middleware":{"Regexp": "^http://[^/]*(.*)$", "Replacement": "https://example.com$1", "Redirect": true}}'
```

**Comment by schmitch** [comment]
@porjo yeah thanks, that works :+1: 

**OFFTOPIC:**
Also I think that sentence is just untrue:

    It's not general-purpose reverse proxy - it is focused on microservices and API use-cases.

Especially since http://blog.vulcanproxy.com/vulcand-news-scoped-listeners/
There are a lot of things that makes vulcand a good reverse proxy for a lot of things and not only microservices.

The only thing which is yet missing would be host grouping i.e. that you could group www.foo.bar and foo.bar as the same host or foo.bar and bar.foo, so that you don't need to configure them twice. 
Yet I don't think that this has a good microservices use case so i don't propose

**Comment by porjo** [comment]
@schmitch you could use a `HostRegexp` to match on multiple hostnames e.g. `www.foo.bar` and `foo.bar`:
```json
etcdctl set /vulcand/frontends/foo/frontend \
'{"Type": "http", "BackendId": "bar", "Route": "HostRegexp(`^(www\\.)?foo\\.bar$`) && PathRegexp(`.*`)"}'
```

Admittedly using regex for this will be inefficient when dealing with a lot of entries.

**Comment by philk** [comment]
I feel like I must be missing something. I tried both

Documentation:
```
vctl rewrite upsert -f f1 -id r1 --regexp="^http://localhost/(.*)" --replacement='https://localhost/$1' --redirect
```

From @porjo above:
```
vctl rewrite upsert -f app --id r1 --regexp='^http://[^/]*(.*)$' --replacement='https://example.com$1' --redirect
```

I get a redirect loop either way. If I just curl the https site directly I get the same 302 and Location as if I curl the http site. Any ideas what I might be missing? Using the mailgun/vulcand:v0.8.0-beta.3 docker container, though I tried it from a container built off master and everything was the same.
